### DISCUSSION TITLE
docs: Add suggestions about getInitialState on persist

### DISCUSSION THREAD

**Discussion by Cadienvan** [pmndrs/zustand#2376] [Bug report]
The `persist` middleware alters the behavior of `getInitialState` because of internal reasons mentioned in #2277.
When following the documentation [here](https://github.com/pmndrs/zustand/blob/16b0ebf1b12e2a9eef88565810185ac7b0c9d2a8/docs/guides/testing.md?plain=1#L82) while having a persisted state accompanied by a non-persisted one, the suggestion about getInitialState in the vite's `afterEach` lifecycle method isn't applicable.

Here's an example:
<img width="520" alt="Screenshot 2024-02-29 alle 14 51 44" src="https://github.com/pmndrs/zustand/assets/5492927/d12a9f05-6620-46f1-835f-4406b74ab22b">

Our service slice is persisted, while the remaining 3 aren't, so our `getInitialState` will only contain the `createServiceSlice` slice of the state.

I'd love to mention it in a PR if you have any idea about how to properly handle this.

**Comment by dai-shi** [comment]
@dbritto-dev ?

**Reply by dbritto-dev** [reply]
@Cadienvan would you mind sharing us a minimal repro on stackblitz?

**Reply by Cadienvan** [reply]
I'll do as soon as I can, in the meantime, I'll give some context info:

A Slice:
<img width="413" alt="Screenshot 2024-02-29 alle 15 38 16" src="https://github.com/pmndrs/zustand/assets/5492927/4cc2e779-0973-4769-a57b-6ed0e62e7a56">

Another Slice:
<img width="696" alt="Screenshot 2024-02-29 alle 15 38 35" src="https://github.com/pmndrs/zustand/assets/5492927/1eed0c74-3072-4fcf-8d3c-90d281ad6df5">

Merged together in a single store:
<img width="528" alt="Screenshot 2024-02-29 alle 15 40 23" src="https://github.com/pmndrs/zustand/assets/5492927/4a67cabc-190b-45b6-bc58-1314d104c582">


It is sufficient to have a persisted store (First one) and a non-persisted store (The second one) together and follow the testing tutorial and, as soon as you `console.log(store.getInitialState())`, only the first slice properties and methods will be available.

This is because the `persist` middleware rewrites the `getInitialState` method.

**Reply by dbritto-dev** [reply]
@Cadienvan I was able to see the issue, you shouldn't use persist on a slice. You should always use persist at store level not at slice level.

If you want to persist a slice you should use persist at store level and partialize to do that

**Reply by Cadienvan** [reply]
That's what I imagined, in fact I agree with the proposed resolution, I'd just mention it in the documentation.

**Reply by dbritto-dev** [reply]
@Cadienvan actually any middleware works at store level but people use to misunderstand that and try to use it on slices

**Reply by Cadienvan** [reply]
@dbritto-dev so do you think adding a line [here](https://github.com/pmndrs/zustand/blob/main/docs/guides/slices-pattern.md) mentioning middlewares should not be used at the slice level isn't necessary?

Of course, mine is only a suggestion but I'd love to be of help being in a team which had such an issue and didn't find info about it.

**Reply by dbritto-dev** [reply]
@Cadienvan no I don't, I'm down with you we shall add a note to make it more clear

**Reply by dbritto-dev** [reply]
@Cadienvan let's create a PR I would be happy to review it
